[iOS] Add support for rendering native selection UI inside scrolled content

whsieh opened the pull request:
#### 9549253a0e1f4ce655aedfa7e140264cae70e298
<pre>
[iOS] Add support for rendering native selection UI inside scrolled content
<a href="https://bugs.webkit.org/show_bug.cgi?id=280591">https://bugs.webkit.org/show_bug.cgi?id=280591</a>

Reviewed by Richard Robinson.

Work towards refactoring selection rendering on iOS, such that native selection UI (e.g. highlight,
caret, etc.) isn&apos;t limited to being installed in `WKContentView`.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:

Add a new runtime preference to make it easier to switch between the new and old selection behavior.

* Source/WebCore/rendering/RenderLayerScrollableArea.h:
* Source/WebKit/Shared/EditorState.h:
* Source/WebKit/Shared/EditorState.serialization.in:

Plumb the `ScrollingNodeID` of the scrollable area containing the selection through `EditorState`,
into the UI process.

* Source/WebKit/UIProcess/ios/WKBaseScrollView.h:
* Source/WebKit/UIProcess/ios/WKBaseScrollView.mm:
(-[WKBaseScrollView scrolledContentView]):

Add a helper property to grab the scrolled content view for the scroll view (i.e. a non-empty
`WKCompositingView` or `WKContentView` under a child or main scroll view, respectively).

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView selectionHonorsOverflowScrolling]):
(-[WKContentView _updateChangedSelection:]):

Call `-prepareToMoveSelectionContainer:` when the selection changes visually (see below).

(-[WKContentView _selectionContainerViewAboveText]):
(-[WKContentView selectionContainerView]):

Implement `-_selectionContainerViewAboveText` and `-selectionContainerView` (both currently SPI) and
return the scrolled contents corresponding to the scroller that currently contains the selection.
Also, add a FIXME to adopt the public API (when it becomes available) and leave behind a call to
`RELEASE_ASSERT_ASYNC_TEXT_INTERACTIONS_DISABLED()` after adoption.

(-[WKContentView _selectionContainerViewInternal]):
* Source/WebKit/UIProcess/ios/WKTextInteractionWrapper.h:
* Source/WebKit/UIProcess/ios/WKTextInteractionWrapper.mm:
(-[WKTextInteractionWrapper textSelectionDisplayInteraction]):
(-[WKTextInteractionWrapper prepareToMoveSelectionContainer:]):

(Indirectly) tell the text selection display interaction to remove and reparent all managed
selection UI views when the selection container view changes. When the new behavior is disabled
(i.e. on shipping iOS) this is going to always be `WKContentView`; however, it may change between a
`WKCompositingView` and the content view when the flag is set, depending on whether the selection is
inside of a scrollable region.

(-[WKTextInteractionWrapper activateSelection]):
(-[WKTextInteractionWrapper deactivateSelection]):
(-[WKTextInteractionWrapper willStartScrollingOverflow]):
(-[WKTextInteractionWrapper didEndScrollingOverflow]):

Don&apos;t hide and reveal the selection in the case where the new behavior is enabled. This is no longer
necessary, since the selection will now simply scroll along with the content, instead of appearing
in the wrong place.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::updatePreferences):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::computeSelectionClipRect const):

Make this additionally compute the enclosing scrolling node ID, along with the selection clip rect.
Once the feature flag is enabled and this behavior becomes the default, this should be renamed to
something like `computeEnclosingNodeIDForSelection` (and the selection clip rect computation should
be moved back out to the call site).

Note that we don&apos;t set the `selectionClipRect` when selection views honor overflow scrolling,
because the selection will be clipped by overflow scrolling with or without the rect.

Canonical link: <a href="https://commits.webkit.org/284461@main">https://commits.webkit.org/284461@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9295d28bec7fa84048f68ec01c60876b78833922

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69467 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48867 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/22140 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/73550 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/20625 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/71584 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56668 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20476 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/55246 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13705 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72533 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44577 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59976 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35725 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41244 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17406 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/19002 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/62583 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63185 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17752 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/75261 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/68713 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/13449 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16973 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/62914 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/13489 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60059 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62820 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15419 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10842 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4459 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/90495 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44671 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/16057 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45745 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46940 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45486 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -13598,6 +13603,41 @@ - (void)_closeCurrentTypingCommand
 
+- (UIView *)_selectionContainerViewAboveText
+{
+    // FIXME: Once -selectionContainerView (or an equivalent) is available as API, we should adopt that and
```
is there a bugzilla bug or radar number you can associate with this?

whsieh replied:
Not at the moment. I filed [280623](https://bugs.webkit.org/show_bug.cgi?id=280623) to track overall enablement, and [280624](https://bugs.webkit.org/show_bug.cgi?id=280624) to track adoption of the requisite API.

rr-codes commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -13598,6 +13603,41 @@ - (void)_closeCurrentTypingCommand
+    if (!scrollingNodeID)
+        return self;
+
+    auto* coordinator = downcast<WebKit::RemoteScrollingCoordinatorProxyIOS>(_page->scrollingCoordinatorProxy());
```
should this be some sort of smart pointer instead?

whsieh replied:
Sort of — it implements `CanMakeWeakPtr` so we can at least hold a weak reference, though I think we'd ideally make this a `CheckedPtr` since the coordinator should never be destroyed as a result of this logic.

I'll make this a weak pointer for now.

whsieh committed:
[iOS] Add support for rendering native selection UI inside scrolled content
https://bugs.webkit.org/show_bug.cgi?id=280591

Reviewed by Richard Robinson.

Work towards refactoring selection rendering on iOS, such that native selection UI (e.g. highlight,
caret, etc.) isn't limited to being installed in `WKContentView`.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:

Add a new runtime preference to make it easier to switch between the new and old selection behavior.

* Source/WebCore/rendering/RenderLayerScrollableArea.h:
* Source/WebKit/Shared/EditorState.h:
* Source/WebKit/Shared/EditorState.serialization.in:

Plumb the `ScrollingNodeID` of the scrollable area containing the selection through `EditorState`,
into the UI process.

* Source/WebKit/UIProcess/ios/WKBaseScrollView.h:
* Source/WebKit/UIProcess/ios/WKBaseScrollView.mm:
(-[WKBaseScrollView scrolledContentView]):

Add a helper property to grab the scrolled content view for the scroll view (i.e. a non-empty
`WKCompositingView` or `WKContentView` under a child or main scroll view, respectively).

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView selectionHonorsOverflowScrolling]):
(-[WKContentView _updateChangedSelection:]):

Call `-prepareToMoveSelectionContainer:` when the selection changes visually (see below).

(-[WKContentView _selectionContainerViewAboveText]):
(-[WKContentView selectionContainerView]):

Implement `-_selectionContainerViewAboveText` and `-selectionContainerView` (both currently SPI) and
return the scrolled contents corresponding to the scroller that currently contains the selection.
Also, add a FIXME to adopt the public API (when it becomes available) and leave behind a call to
`RELEASE_ASSERT_ASYNC_TEXT_INTERACTIONS_DISABLED()` after adoption.

(-[WKContentView _selectionContainerViewInternal]):
* Source/WebKit/UIProcess/ios/WKTextInteractionWrapper.h:
* Source/WebKit/UIProcess/ios/WKTextInteractionWrapper.mm:
(-[WKTextInteractionWrapper textSelectionDisplayInteraction]):
(-[WKTextInteractionWrapper prepareToMoveSelectionContainer:]):

(Indirectly) tell the text selection display interaction to remove and reparent all managed
selection UI views when the selection container view changes. When the new behavior is disabled
(i.e. on shipping iOS) this is going to always be `WKContentView`; however, it may change between a
`WKCompositingView` and the content view when the flag is set, depending on whether the selection is
inside of a scrollable region.

(-[WKTextInteractionWrapper activateSelection]):
(-[WKTextInteractionWrapper deactivateSelection]):
(-[WKTextInteractionWrapper willStartScrollingOverflow]):
(-[WKTextInteractionWrapper didEndScrollingOverflow]):

Don't hide and reveal the selection in the case where the new behavior is enabled. This is no longer
necessary, since the selection will now simply scroll along with the content, instead of appearing
in the wrong place.

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::updatePreferences):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::computeSelectionClipRect const):

Make this additionally compute the enclosing scrolling node ID, along with the selection clip rect.
Once the feature flag is enabled and this behavior becomes the default, this should be renamed to
something like `computeEnclosingNodeIDForSelection` (and the selection clip rect computation should
be moved back out to the call site).

Note that we don't set the `selectionClipRect` when selection views honor overflow scrolling,
because the selection will be clipped by overflow scrolling with or without the rect.

Canonical link: https://commits.webkit.org/284461@main
